aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+    value is ignored.
+
+    The items in the returned vector are in increasing order.
+  */
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  std::vector<iscalar_t> offsets(nnz, 0);
+  std::vector<iscalar_t> strides(ndim, 1);
+  iscalar_t* indices_data_base = indices.data_ptr<iscalar_t>();
+
+  if (ndim > 1) {
+    for (int64_t i=ndim - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * sizes[i + 1];
+    }
+  }
+
+  for (int64_t j=0; j < ndim; j++) {
+    if (j != dim) {
+      for (int64_t i=0; i < nnz; i++) {
+        offsets[i] += strides[j] * indices_data_base[j * nnz + i];
+      }
+    }
+  }
+
+  return offsets;
+}
+
+template <typename iscalar_t>
+std::vector<iscalar_t> get_dense_offsets(iscalar_t &mx, const std::vector<iscalar_t>& offsets) {
+  /*
+    Return the dense set of offsets:
+
+      If
+        pool = get_dense_offsets(mx, offsets)
+      then
+        pool.size() == offsets.size()
+        pool[i] == pool[j] iff offsets[i] == offsets[j]
+        set(pool) == set(range(mx))
+  */
+  auto n = offsets.size();
+  std::vector<iscalar_t> pool(n, 0);
+  std::map<iscalar_t, iscalar_t> i2p;
+  for (int64_t i=0; i < n; i++) {
+    auto c = offsets[i];
+    auto it = i2p.find(c);
+    iscalar_t p = i2p.size();
+    if (it == i2p.end()) {
+      i2p.emplace(std::make_pair(c, p));
+    } else {
+      p = it->second;
+    }
+    pool[i] = p;
+  }
+  mx = i2p.size();
+  return pool;
+}
+
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
+void cpu_sparse_coo_softmax(Tensor output, const Tensor& input, const int64_t dim) {
+  /*
+    See test/test_sparse.py:test_softmax:sparse_softmax for the Python
+    prototype of the sparse softmax algorithm that this implementation
+    is based on.
+   */
+  auto sparse_dim = input.sparse_dim();
+  auto indices = input._indices().contiguous();
+  auto values = input._values().contiguous();
+  auto out_values = output._values();
+  auto out_indices = output._indices();
+  out_values.resize_as_(values);
+  out_indices.resize_as_(indices);
+  out_indices.copy_(indices);
+
+  if (dim >= sparse_dim) {
+    if (LogSoftMax) {
+      auto new_values = log_softmax_cpu(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    } else {
+      auto new_values = softmax_cpu(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    }
+    return;
+  }
+
+  auto nnz = values.size(0);
+  scalar_t* values_data_base = values.data_ptr<scalar_t>();
+  iscalar_t* indices_data_base = indices.data_ptr<iscalar_t>();
+  scalar_t* out_values_data_base = out_values.data_ptr<scalar_t>();
+  auto sizes = input.sizes().vec();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Compute pool indices */
+  std::vector<iscalar_t> offsets = get_offsets<iscalar_t>(indices, sizes, dim);
+  iscalar_t mx_p;
+  std::vector<iscalar_t> pool = get_dense_offsets<iscalar_t>(mx_p, offsets);
```

Comment:
nit: Could you please insert more explanation about this `pool` mapping?

Reply:
Yes, the sparse/SoftMax.cpp now contains the derivation of the sparse softmax algorithm with an example.